### PR TITLE
Added chapter on advanced generics

### PR DESCRIPTION
**Pull request by gusthoff**

### CODE REVIEW

File: courses/advanced-ada/chapters/generics.rst [hunk 1]
```diff
@@ -408,3 +392,1511 @@ two:
...
+In this example, we declare the data container :ada:`C` and bind it to
+two instantiations of the :ada:`Display` procedure:
+
+- :ada:`Display_First_Name_Age`, which displays the first name and age of
+  each person from the database;
+
+- :ada:`Display_Name_Birthday`, which displays the full name and birthday
+  of each person.
+
+Formal access types
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: what you describe below is not formal access types (see http://www.ada-auth.org/standards/rm12_w_tc1/html/RM-12-5-4.html) but formal object parameters of access (to subprogram) type. I'm not sure this section adds anything useful.

File: courses/advanced-ada/chapters/generics.rst [hunk 2]
```diff
@@ -408,3 +392,1511 @@ two:
...
+       Put_Line ("Avg: " & Float'Image (Avg));
+
+    end Show_Average;
+
+In this example, we declare the array :ada:`A` with elements of both
+:ada:`T` and :ada:`T2` types. After initializing the elements of :ada:`A`,
+we call the :ada:`Average` function from :ada:`Ops`, an instance of the
+generic package :ada:`Float_Interface_Pkg.Ops`.
+
+Discussion: Generic interfaces vs. other approaches
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: This section is empty, is it ok or do you want to comment it out for now?

File: courses/advanced-ada/chapters/generics.rst [hunk 3]
```diff
@@ -408,3 +392,1511 @@ two:
...
+Alternatively, we could make use of the :ada:`Float_Types` package
+directly in the generic package. For example:
+
+.. code:: ada
+
+    with Float_Types; use Float_Types;
+
+    generic
+       type F is new My_Float;
+       with function "+" (A : F; B : F) return F is <>;
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: I'm not sure you still want to pass in the "+" as formal parameter in this case

File: courses/advanced-ada/chapters/generics.rst [hunk 4]
```diff
@@ -408,3 +392,1511 @@ two:
...
+         with Size => Ovhd_Depth;
+
+       --  Ensure that 'First and 'Last have at least double amount
+       --  of bits as the original type
+       pragma Assert (Ovhd_Fixed'First <=
+                      Ovhd_Fixed (-2.0 ** (F'Size - 1)));
+       pragma Assert (Ovhd_Fixed'Last  >=
+                      Ovhd_Fixed (2.0 ** (F'Size - 1) - Ovhd_Delta));
+
+       --  Ensure that 'Size is has at least twice as many bits as
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: typo: "is has"

File: courses/advanced-ada/chapters/generics.rst [hunk 5]
```diff
@@ -408,3 +392,1511 @@ two:
...
+
+       Ovhd_Depth : constant Positive := 64;
+       Ovhd_Bits  : constant := 32;
+       Ovhd_Delta : constant := 2.0 ** Ovhd_Bits / 2.0 ** (Ovhd_Depth - 1);
+
+       type Ovhd_Fixed is delta Ovhd_Delta range
+         -2.0 ** Ovhd_Bits .. 2.0 ** Ovhd_Bits - Ovhd_Delta
+         with Size => Ovhd_Depth;
+
+       --  Ensure that 'First and 'Last have at least double amount
```
Comment (yannickmoy) [inline courses/advanced-ada/chapters/generics.rst]: I don't understand how the assertions that follow ensure what you need. Why not simply

pragma Assert (Ovhd_Fixed'First <= 2 * Ovhd_Fixed(F'First));
pragma Assert (Ovhd_Fixed'Last >= 2 * Ovhd_Fixed(F'Last));

Comment (gusthoff) [reply courses/advanced-ada/chapters/generics.rst]: Yes, you're right. I've got those from an implementation for a saturated multiplication. For a saturated addition, it's not really needed. I'll change it later.

**Pull request merged**